### PR TITLE
Create login route

### PR DESCRIPTION
**Pull request by khaled-samy**
### Solves  #16 
- created login validation
- created login controller

### CODE REVIEW

File: server/controllers/auth/login.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const { rowCount: isEmailTaken, rows } = await checkEmailTakenQuery({ destination, email });
+    if (!isEmailTaken) throw new CustomError('Your email is incorrect', 401);
+    const { password: hashedPassword } = rows[0];
+    const isPasswordMatch = await compare(password, hashedPassword);
+    if (!isPasswordMatch) throw new CustomError('Your password is incorrect', 401);
+    const token = await signToken({ email, role });
+    const user = rows[0];
+    res
+      .cookie('token', token)
+      .json({ data: user, message: 'User Log in Successfully' });
```
Comment (zaher-aa) [inline server/controllers/auth/login.ts:23]: Status code for creating a new token is 201
```suggestion
    res
      .status(201)
      .cookie('token', token)
      .json({ data: user, message: 'User Log in Successfully' });
```

File: server/controllers/auth/login.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      role, email, password,
+    } = req.body;
+    await loginSchema.validate({ email, password, role });
+    const destination = role === 'student' ? 'students' : 'teachers';
+    const { rowCount: isEmailTaken, rows } = await checkEmailTakenQuery({ destination, email });
+    if (!isEmailTaken) throw new CustomError('Your email is incorrect', 401);
+    const { password: hashedPassword } = rows[0];
+    const isPasswordMatch = await compare(password, hashedPassword);
+    if (!isPasswordMatch) throw new CustomError('Your password is incorrect', 401);
+    const token = await signToken({ email, role });
```
Comment (zaher-aa) [inline server/controllers/auth/login.ts:19]: The token should have user's id, name and role
```suggestion
    const token = await signToken({ userId, username, role });
```

**Review by zaher-aa — CHANGES_REQUESTED** [review]

File: server/controllers/auth/login.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    const destination = role === 'student' ? 'students' : 'teachers';
+    const { rowCount: isEmailTaken, rows } = await checkEmailTakenQuery({ destination, email });
+    if (!isEmailTaken) throw new CustomError('Your email is incorrect', 401);
+    const { password: hashedPassword } = rows[0];
+    const isPasswordMatch = await compare(password, hashedPassword);
+    if (!isPasswordMatch) throw new CustomError('Your password is incorrect', 401);
+    const token = await signToken({ userId, username, role });
+    const user = rows[0];
+    res
+      .status(201)
```
Comment (amjed-98) [inline server/controllers/auth/login.ts:22]: status should be 200

**Review by amjed-98 — CHANGES_REQUESTED** [review]
status should be 200
image

File: server/controllers/auth/login.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      role, email, password,
+    } = await loginSchema.validate(req.body, { abortEarly: false });
+    const destination = role === 'student' ? 'students' : 'teachers';
+    const { rowCount: isEmailTaken, rows } = await checkEmailTakenQuery({ destination, email });
+    if (!isEmailTaken) throw new CustomError('Your email is incorrect', 401);
+    const {
+      password: hashedPassword, username, id,
+    } = rows[0];
+    const isPasswordMatch = await compare(password, hashedPassword);
+    if (!isPasswordMatch) throw new CustomError('Your password is incorrect', 401);
```
Comment (Mu7ammadAbed) [inline server/controllers/auth/login.ts:19]:
```suggestion
    if (!isPasswordMatch) throw new CustomError('Incorrect email or password', 401);
```

File: server/controllers/auth/login.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import { CustomError } from '../../errors';
+
+export default async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const {
+      role, email, password,
+    } = await loginSchema.validate(req.body, { abortEarly: false });
+    const destination = role === 'student' ? 'students' : 'teachers';
+    const { rowCount: isEmailTaken, rows } = await checkEmailTakenQuery({ destination, email });
+    if (!isEmailTaken) throw new CustomError('Your email is incorrect', 401);
```
Comment (Mu7ammadAbed) [inline server/controllers/auth/login.ts:14]:
```suggestion
    if (!isEmailTaken) throw new CustomError('Incorrect email or password', 401);
```

File: server/controllers/auth/login.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    const {
+      password: hashedPassword, username, id,
+    } = rows[0];
+    const isPasswordMatch = await compare(password, hashedPassword);
+    if (!isPasswordMatch) throw new CustomError('Your password is incorrect', 401);
+    const token = await signToken({ id, username, role });
+
+    res
+      .status(200)
+      .cookie('token', token)
```
Comment (Mu7ammadAbed) [inline server/controllers/auth/login.ts:24]:
```suggestion
      .cookie('token', token, {
        httpOnly: true,
        secure: process.env.NODE_ENV === 'production',
      })
```

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by zaher-aa — APPROVED** [review]

**Review by amjed-98 — APPROVED** [review]
great job

**Pull request merged**